### ISSUE TITLE
Create a divergence-free 3D velocity field to be used in examples

### ISSUE THREAD

**Issue by zhenwu0728** [JuliaOcean/PlanktonIndividuals.jl#19]
@gaelforget You've got a 2D stream function to generate a divergence-free velocity field. And we need to extend it to 3D with periodic horizontal boundary conditions and bounded vertical boundary conditions. I haven't figured out a way to do that. It would be great if you can help.

**Comment by christophernhill** [comment]
@zhenwu0728 two ideas

1. we sometimes define a 3d flow by summing two or more 2d stream functions in different planes (xy, xz, yz). The gradient of that will then give a non-divergent flow.

2. you can solve a Poisson equation for a potential whose gradient when subtracted from the flow will make the flow non-divergent. This is useful for flows derived from obs for example. There is some Julia code here https://github.com/CliMA/Oceananigans.jl/blob/master/test/test_preconditioned_conjugate_gradient_solver.jl that could be adapted for that.

**Comment by gaelforget** [comment]
> @gaelforget You've got a 2D stream function to generate a divergence-free velocity field. And we need to extend it to 3D with periodic horizontal boundary conditions and bounded vertical boundary conditions. I haven't figured out a way to do that. It would be great if you can help.

Hi @zhenwu0728 

Happy to help with this as discussed yesterday

**Comment by gaelforget** [comment]
> @zhenwu0728 two ideas
> 
> 1. we sometimes define a 3d flow by summing two or more 2d stream functions in different planes (xy, xz, yz). The gradient of that will then give a non-divergent flow.
> 2. you can solve a Poisson equation for a potential whose gradient when subtracted from the flow will make the flow non-divergent. This is useful for flows derived from obs for example. There is some Julia code here https://github.com/CliMA/Oceananigans.jl/blob/master/test/test_preconditioned_conjugate_gradient_solver.jl that could be adapted for that.



> @zhenwu0728 two ideas
> 
> 1. we sometimes define a 3d flow by summing two or more 2d stream functions in different planes (xy, xz, yz). The gradient of that will then give a non-divergent flow.
> 2. you can solve a Poisson equation for a potential whose gradient when subtracted from the flow will make the flow non-divergent. This is useful for flows derived from obs for example. There is some Julia code here https://github.com/CliMA/Oceananigans.jl/blob/master/test/test_preconditioned_conjugate_gradient_solver.jl that could be adapted for that.

Thanks @christophernhill , as you said it's a pretty standard thing to do -- we were talking about these options with @zhenwu0728 when I offered to help

FYI, you will find a general 2D implementation for this in MeshArrays.jl which readily covers cases with coast lines, 3D topography, and uneven grids like cube spheres and such, which (as you of course know) are common in models and data. There is an example in https://github.com/JuliaClimate/GlobalOceanNotebooks (`OceanTransports/05_streamfunction.ipynb`) in case that's useful to you guys

**Comment by zhenwu0728** [comment]
I've played with 2D stream functions in both periodic boundary conditions and bounded boundary conditions. The periodic one works perfectly, but the bounded one doesn't work really well.

The reason I opened this issue is that when I advect particles based on the velocity fields generated by Oceananigans.jl ([https://github.com/zhenwu0728/TurbulentBlooms.jl/blob/master/diel_heat_flux.jl](url), mixing happens at the top 50 meters due to the diel heat flux). The particles tend to accumulate in the surface layer which is abnormal (see figure below, randomly generated particles with mixing in the top 50 m). 

As @gaelforget pointed out, this might be related to the divergence of the velocity field. So I'm wondering if there's a way to correct it.
<img width="200" alt="Screen Shot 2021-03-05 at 12 21 49 PM" src="https://user-images.githubusercontent.com/46291075/110150574-6ad39f00-7dad-11eb-97d4-196579465140.png">

**Comment by zhenwu0728** [comment]
@gaelforget as discussed this afternoon, I've set up the [vertical 2D example](https://github.com/JuliaOcean/PlanktonIndividuals.jl/blob/master/examples/vertical_2D_example.jl). But I found that the first and last layers of u are zeros in z-direction but not zeros in x-direction (see plot below). Similarly, the first and last layers of w are zeros in x-direction but not zeros in z-direction. 
I think that's not the flow field we discussed this afternoon, a 90 degree rotation might be needed, I think. Correct me if I'm wrong.
<img width="400" alt="Screen Shot 2021-03-17 at 10 20 46 PM" src="https://user-images.githubusercontent.com/46291075/111563819-0fbb8800-876f-11eb-90a3-aa8a7204dfec.png">
